### PR TITLE
Add Support for NET 5 to Build- Take 2

### PR DESCRIPTION
**Pull request by eaba**
This PR fulfills https://github.com/akkadotnet/akka.net/issues/4667 

- [x] Initial commit to test everything before making any changes [[de4787b](https://github.com/akkadotnet/akka.net/pull/4683/commits/de4787be7c4abf56243ef555d85ce877e0cfe71e)]
- [x] Bump Windows Image version [[d3eb2b7](https://github.com/akkadotnet/akka.net/pull/4683/commits/d3eb2b71d24197aff06b0d42b9de4a1b267470e4)]
- [x] Bump Ubuntu Image version [[07091de](https://github.com/akkadotnet/akka.net/pull/4683/commits/07091de73eadd33920f40c059dc4f00cb09e08be)]
- [x] Install .NET 5 SDK and .NET Core 3.1 Runtime [[0ccd4cd](https://github.com/akkadotnet/akka.net/pull/4683/commits/0ccd4cd45ca45364b9de59a3f70e87c084b07a82) [c6ccb6e](https://github.com/akkadotnet/akka.net/pull/4683/commits/c6ccb6ece41721a016a79562f8a4562388626301)]
- [x] Add .NET 5 framework target to all Test projects [[8f727a9](https://github.com/akkadotnet/akka.net/pull/4683/commits/8f727a94875605610483888de68d2a6a582b8e49) [2f4daf7](https://github.com/akkadotnet/akka.net/pull/4683/commits/2f4daf7f28c9291162ddfd6ee6a0a6ee9eadbfe6) [e7ac08f](https://github.com/akkadotnet/akka.net/pull/4683/commits/e7ac08f14b244fadd957d6723ade5fb0a2e2616c) ]
- [x] Add Azure pipeline Test for .NET5.0 [[644e974](https://github.com/akkadotnet/akka.net/pull/4683/commits/644e9740f7ebc5e7ba540aad83f02162cacd61bc) [497e0d5](https://github.com/akkadotnet/akka.net/pull/4683/commits/497e0d5f532a3f762a650004ca1303281a9a83f2) [522f21b](https://github.com/akkadotnet/akka.net/pull/4683/commits/522f21ba0c5c031ea3795f88bf80ef1eec52c0ea)]
- [x] Bump Incrementalist to newer version with .NET 5 support
- [x] Update MultiNodeTestRunner for .NET5 platform [[a97ad04](https://github.com/akkadotnet/akka.net/pull/4683/commits/a97ad0411130daf5a235763a6e23a86782ede6ec)]

### CODE REVIEW

**Comment by eaba** [comment]
PR Validation (Windows Build)  Failure
````
2020-12-23T12:37:03.5632452Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(15,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5635634Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(20,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5638463Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(16,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5642108Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5645242Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5649753Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(26,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5654092Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(26,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5656999Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5660362Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:37:03.5663330Z     9 Warning(s)
2020-12-23T12:37:03.5664433Z     9 Error(s)````

**Comment by eaba** [comment]
PR Validation (NuGet Pack) Failure
````
2020-12-23T12:39:14.9945043Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(15,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9949422Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(20,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9952108Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(16,7): error CS0246: The type or namespace name 'NUnit' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9955368Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9959862Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaIdentityProcessorVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9962673Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(26,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9965824Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaPublisherVerification.cs(26,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9970188Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixtureAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9972848Z D:\a\1\s\src\core\Akka.Streams.Tests.TCK\AkkaSubscriberVerification.cs(21,6): error CS0246: The type or namespace name 'TestFixture' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\core\Akka.Streams.Tests.TCK\Akka.Streams.Tests.TCK.csproj]
2020-12-23T12:39:14.9974461Z     9 Warning(s)
2020-12-23T12:39:14.9975819Z     9 Error(s)
````

**Comment by eaba** [comment]
NUnit support for .net5.0 was merged recently and not released yet!

**Comment by Aaronontheweb** [comment]
> NUnit support for .net5.0 was merged recently and not released yet!

Think it's possible to use the test filters inside FAKE to skip .NET 5 execution of the Reactive.Streams.TCK? That's the only library that depends on NUnit. We might also want to spend a pull request to them if they need to update their NUnit version: https://github.com/reactive-streams/reactive-streams-dotnet

**Comment by eaba** [comment]
The problem is that it fails to build.....

**Comment by Aaronontheweb** [comment]
> The problem is that it fails to build.....

Then don't build that library for .NET 5 - and leave a comment inside the `.csproj` file explaining why

**Comment by eaba** [comment]
> 
> 
> > The problem is that it fails to build.....
> 
> Then don't build that library for .NET 5 - and leave a comment inside the `.csproj` file explaining why

Okay

**Comment by eaba** [comment]
image

**Comment by Aaronontheweb** [comment]
@eaba looks like you found a DData bug - can you file an issue and patch it?

Looks like Linux is failing due to Incrementalist - we should get a new release out there today and see if those upgrades you've been working on will fix the issue. I'll take a look at that now.

**Comment by Aaronontheweb** [comment]
This is looking very promising though!

**Comment by Aaronontheweb** [comment]
Re-running this with Incrementalist v0.4.0 added

**Comment by Aaronontheweb** [comment]
```
if [ ! -f "$INCREMENTALIST_EXE" ]; then
    "$SCRIPT_DIR/.dotnet/dotnet" tool install Incrementalist.Cmd --version $INCREMENTALIST_VERSION --tool-path "$INCREMENTALIST_DIR"
    if [ $? -ne 0 ]; then
        echo "Incrementalist already installed."
    fi
```

This is why Incrementalist is failing on Linux - now that we don't install .NET Core locally into the `.dotnet` folder Incrementalist can't be installed. You should mirror what we do in `build.ps1` - that should work here I think.

**Comment by eaba** [comment]
````
Async error during test execution: OnSubscribe should be called prior to OnComplete always
at Reactive.Streams.TCK.TestEnvironment.VerifyNoAsyncErrorsNoDelay()
at Reactive.Streams.TCK.PublisherVerification`1.<Required_spec109_mustIssueOnSubscribeForNonNullSubscriber>b__37_0(IPublisher`1 publisher)
at Reactive.Streams.TCK.PublisherVerification`1.ActivePublisherTest(Int64 elements, Boolean completionSignalRequired, Action`1 run)
at Reactive.Streams.TCK.PublisherVerification`1.Required_spec109_mustIssueOnSubscribeForNonNullSubscriber()
at Reactive.Streams.TCK.IdentityProcessorVerification`1.Required_spec109_mustIssueOnSubscribeForNonNullSubscriber()

````

**Comment by eaba** [comment]
image

**Comment by eaba** [comment]
image

**Comment by eaba** [comment]
image

**Comment by Aaronontheweb** [comment]
@eaba those look like legit bugs created with these changes - you should debug them locally. My guess is it's related to `ConfigurationManager` in .NET 5

**Comment by eaba** [comment]
image
image
This are likely to pass in another run!

**Comment by Aaronontheweb** [comment]
@eaba let me see if those were racy specs - going to re-run CI

**Comment by Aaronontheweb** [comment]
@akkadotnet/core so it looks like this is going to add:

1. .NET 5 tests on Windows
2. .NET 5 tests on Linux
3. .NET 5 MNTR tests

To each build - so that's some more load time included. I think it's probably a good idea from a "completeness" perspective to test everything on these platforms but it's going to eat up a lot more time per-build especially when anyone touches `Akka.csproj`, which will launch a full build. Should we look at maybe trimming .NET Core 3.1 out of the build cycle relatively soon in the future to reduce build times?

File: build.fsx [hunk 1]
```diff
@@ -279,6 +284,32 @@ Target "RunTestsNetCore" (fun _ ->
         projects |> Seq.iter (runSingleProject)
 )
 
+Target "RunTestsNet" (fun _ ->
```
Comment (Aaronontheweb) [inline build.fsx:287]: Changes here look great - nice work

File: build.ps1 [hunk 2]
```diff
@@ -48,49 +45,6 @@ if (!(Test-Path $ToolPath)) {
     New-Item -Path $ToolPath -Type directory | out-null
 }
 
-###########################################################################
```
Comment (Aaronontheweb) [inline build.ps1:51]: Should resolve https://github.com/akkadotnet/akka.net/issues/3592 now that we're not downloading and installing `dotnet` now

File: build.sh [hunk 3]
```diff
@@ -44,6 +45,22 @@ if [ ! -d "$TOOLS_DIR" ]; then
   mkdir "$TOOLS_DIR"
 fi
 
+###########################################################################
+# INSTALL .NET 5 SDK
+###########################################################################
+
+echo "Installing .NET 5 SDK..."
```
Comment (Aaronontheweb) [inline build.sh:52]: Do we need to do this? Can't we remove it now that Azure Pipelines handles it for us?

File: src/common.props [hunk 4]
```diff
@@ -16,6 +16,7 @@
     <NBenchVersion>2.0.1</NBenchVersion>
     <ProtobufVersion>3.14.0</ProtobufVersion>
     <NetCoreTestVersion>netcoreapp3.1</NetCoreTestVersion>
+    <NetTestVersion>net5.0</NetTestVersion>
```
Comment (Aaronontheweb) [inline src/common.props:19]: LGTM

File: src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.nuspec.template [hunk 5]
```diff
@@ -11,11 +11,12 @@
     <projectUrl>https://github.com/akkadotnet/akka.net</projectUrl>
     <iconUrl>http://getakka.net/images/AkkaNetLogo.Normal.png</iconUrl>
     <description>Akka.NET Multi-node Test Runner; used for executing tests written with Akka.Remote.TestKit</description>
-    <copyright>Copyright � 2013-2019 .NET Foundation</copyright>
+    <copyright>Copyright � 2013-2020 .NET Foundation</copyright>
```
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.nuspec.template:14]: Nice catch

File: src/core/Akka.MultiNodeTestRunner/Program.cs [hunk 6]
```diff
@@ -309,7 +311,7 @@ static void Main(string[] args)
                                 var nodeRole = nodeTest.Role;
 
 #if CORECLR
-                            if (platform == "netcore")
+                            if (platform == "netcore" || platform == "net5")
```
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner/Program.cs:314]: LGTM

File: src/core/Akka.Streams.Tests.TCK/Akka.Streams.Tests.TCK.csproj [hunk 7]
```diff
@@ -4,6 +4,7 @@
   <PropertyGroup>
     <AssemblyName>Akka.Streams.Tests.TCK</AssemblyName>
     <!-- Reactive.Streams.TCK isn't compatible with .NET Core -->
+    <!--UNit.Framework does not yet support .NET 5 - build will fail-->
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams.Tests.TCK/Akka.Streams.Tests.TCK.csproj:7]: @eaba could you add a Github issue for this so we don't forget?

**Review by Aaronontheweb — APPROVED** [review]
LGTM - great work.

Going to merge this in after 1.4.14 is deployed, just in case.

**Comment by Aaronontheweb** [comment]
Woah... Just noticed that the build time just about doubled in this PR... Went from about 8 minutes to 15? i.e. if you compare to https://github.com/akkadotnet/akka.net/pull/4700

Is this just from adding .NET 5 as a target for our unit test projects @eaba ?

**Comment by eaba** [comment]
> 
> 
> Woah... Just noticed that the build time just about doubled in this PR... Went from about 8 minutes to 15? i.e. if you compare to #4700
> 
> Is this just from adding .NET 5 as a target for our unit test projects @eaba ?

I can't say for sure - the build before the current build took 13m. If each project is [built serially](https://docs.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=vs-2019#target-build-order)  for each target, then maybe we should expect increase in build time.

**Comment by Aaronontheweb** [comment]
@eaba no, it definitely didn't take 13 minutes. I have it clocked in that previous PR at 8. You can see that on every PR that touches `Akka.csproj` can you please investigate? Where's the double time coming from?

**Comment by eaba** [comment]
The installation of the .NET 5 SDK takes about 1m 30s which is over 100% the time it takes to install .NETCore 3.1 SDK (23s from https://github.com/akkadotnet/akka.net/pull/4700)

image

**Comment by eaba** [comment]
Comparing the two PR, package restoration takes twice as much with this PR

image


image

**Comment by Danthar** [comment]
We have seen this on our own production env as well. Builds on Azure Pipelines taking longer because for some reason network activity is delayed. This results in longer download times for nuget packages and anything else that needs to be downloaded. 
This delay scenario can last anywhere from a few hours to a few days. 
Also note that by default max run time is 30 mins for a pipeline (it was in our env anyways). If you go beyond that, builds will be aborted. So if you see something like that happening, odds are we would need to explicitly configure the akka builds to be allowed to run longer.

**Comment by Aaronontheweb** [comment]
> We have seen this on our own production env as well. Builds on Azure Pipelines taking longer because for some reason network activity is delayed. This results in longer download times for nuget packages and anything else that needs to be downloaded.
> This delay scenario can last anywhere from a few hours to a few days.
> Also note that by default max run time is 30 mins for a pipeline (it was in our env anyways). If you go beyond that, builds will be aborted. So if you see something like that happening, odds are we would need to explicitly configure the akka builds to be allowed to run longer.

Think this is related to downloading and installing two different .NET SDKs? We provision that through the pipeline definitions themselves now.

We've had some pipeline stages that take 1hr 15minutes as-is and those appear to still be running just fine.

**Comment by Aaronontheweb** [comment]
Looks like installing .NET 5 does take about 1 minute, and then there is some additional build overhead for the new .NET 5 targets. Oh well - we'll figure out how to trim the time back down eventually. Nice work @eaba

**Pull request merged**